File: TypeScript/schemaAnidado.ts
import { Agent, tool, StructuredOutputError } from "@strands-agents/sdk";
import { z } from "zod";

const gastoSchema = z.object({
  descripcion: z.string().describe("Qué fue el gasto"),
  montoOriginal: z.number(),
  monedaOriginal: z.string(),
  tazaUsada: z.number(),
  montoEnMxn: z.number().describe("Monto convertido a MXN"),
});

const resumenGastosSchema = z.object({
  gastos: z.array(gastoSchema),
  totalMxn: z.number().describe("Suma de todos los gastos convertidos a MXN"),
});

const sumar = tool({
  name: "Sumar_tool",
  description: "Suma n cantidad de números, ej: '1 + 2 + 3'",
  inputSchema: z.object({
    numeros: z.array(z.number()),
  }),
  callback: async ({ numeros }) => {
    try {
      if (!Array.isArray(numeros) || numeros.length === 0) {
        return "Error: no se proporcionaron números para sumar o la lista está vacía.";
      }
      if (!numeros.every(num => typeof num === "number" && !isNaN(num))) {
        return "Error: todos los elementos deben ser números válidos.";
      }
      const resultado = numeros.reduce((acc, curr) => acc + curr, 0);
      return `El resultado de la suma es ${resultado}`;
    } catch (error) {
      return `Error al sumar los números: ${error instanceof Error ? error.message : String(error)}`;
    }
  },
})

const CurrencyConverter = tool({
  name: "currency_converter",
  description: "Convierte una cantidad entre MXN y USD (tasa simulada: 1 USD = 18 MXN). Recibe amount, fromCurrency y toCurrency (usa 'MXN' o 'USD').",
  inputSchema: z.object({
    amount: z.number(),
    fromCurrency: z.enum(["MXN", "USD"]),
    toCurrency: z.enum(["MXN", "USD"]),
  }),
  callback: ({ amount, fromCurrency, toCurrency }) => {
    try {
      if (fromCurrency === toCurrency) {
      return `${amount} ${fromCurrency} equivalen a ${amount} ${toCurrency}`;
    }
    // Tasas relativas a MXN (moneda base)
    const tasasEnMxn: Record<string, number> = { MXN: 1, USD: 18 };
    const amountEnMxn = amount * tasasEnMxn[fromCurrency];
    const resultado = amountEnMxn / tasasEnMxn[toCurrency];
    return `${amount} ${fromCurrency} equivalen a ${resultado.toFixed(2)} ${toCurrency}`;
    } catch (error) {
      return `Error al convertir la moneda: ${error instanceof Error ? error.message : String(error)}`;
    }
  },
});

const agent = new Agent({
  model: "us.anthropic.claude-haiku-4-5-20251001-v1:0",
  systemPrompt: "Eres un asistente de finanzas personales. Usa currency_converter para normalizar cada gasto a MXN y Sumar_tool para hacer sumas",
  tools: [CurrencyConverter, sumar],
  structuredOutputSchema: resumenGastosSchema,
});

const resultado = await agent.invoke("Gasté 500 MXN en comida, 30 USD en transporte y 15.50 USD en café");
const resumen = resultado.structuredOutput as z.infer<typeof resumenGastosSchema>;
resumen.gastos.forEach(g => console.log(`${g.descripcion}: ${g.montoOriginal} ${g.monedaOriginal} = ${g.montoEnMxn.toFixed(2)} MXN`));
console.log(`Total: ${resumen.totalMxn.toFixed(2)} MXN`);

File: Python/schema_anidado.py
from pydantic import BaseModel, Field
from strands import Agent, tool

@tool
def sumar(numeros: list[float]) -> str:
    """Suma n cantidad de números, ej: '1 + 2 + 3'"""
    if not isinstance(numeros, list):
        return "Error: se espera una lista de números."
    if not numeros:
        return "Error: no se proporcionaron números para sumar."
    if not all(isinstance(x, (int, float)) and not isinstance(x, bool) for x in numeros):
        return "Error: todos los elementos deben ser números válidos."
    resultado = sum(numeros)
    return f"El resultado de la suma es {resultado}"

@tool
def currency_converter(amount: float, fromCurrency: str, toCurrency: str) -> str:
    """Convierte MXN a USD y viceversa, ej: '100 USD a MXN'"""
    # Aquí podrías implementar la lógica de conversión de divisas usando una API externa
    # Por simplicidad, vamos a devolver un mensaje simulado
    try:
        if fromCurrency == toCurrency:
            return f"{amount} {fromCurrency} equivalen a {amount} {toCurrency}"

        tazas_en_mx = {"USD": 18, "MXN": 1}
        amount_in_mxn = amount * tazas_en_mx[fromCurrency]
        resultado = amount_in_mxn / tazas_en_mx[toCurrency]
        return f"Convertido {amount} {fromCurrency} a {toCurrency}: {resultado:.2f} {toCurrency}"
    except Exception as e:
        return f"Error al convertir la moneda: {e}"

class Gasto(BaseModel):
    descripcion: str = Field(description="Qué fue el gasto")
    monto_original: float
    moneda_original: str
    monto_en_mxn: float = Field(description="Monto convertido a MXN")

class ResumenGastos(BaseModel):
    gastos: list[Gasto]
    total_mxn: float = Field(description="Suma de todos los gastos convertidos a MXN")

agent = Agent(
    model="us.anthropic.claude-haiku-4-5-20251001-v1:0",
    system_prompt="Eres un asistente de finanzas personales. Usa currency_converter para normalizar cada gasto a MXN y sumar para hacer sumas",
    tools=[currency_converter, sumar],
    structured_output_model=ResumenGastos,
)

resultado = agent("Gasté 500 MXN en comida, 30 USD en transporte y 15.50 USD en café")
resumen = resultado.structured_output

if not isinstance(resumen, ResumenGastos):
    raise TypeError("La respuesta del agente no devolvió un ResumenGastos válido.")

for g in resumen.gastos:
    print(f"{g.descripcion}: {g.monto_original} {g.moneda_original} = {g.monto_en_mxn:.2f} MXN")
print(f"Total: {resumen.total_mxn:.2f} MXN")